packages/react/src/hooks/useDataStore.tsx

```diff
@@ -11,46 +11,103 @@ import {
+ *
+ * @param props object containing params for calling either calling Datastore.query or Datastore.observeQuery
+ * @returns the result of either calling Datastore.query or Datastore.observeQuery for the provided props
  */
-export const useDataStoreCollection = <M extends PersistentModel>({
-  model,
-  criteria,
-  pagination,
-}: DataStoreCollectionProps<M>): DataStoreCollectionResult<M> => {
-  const [result, setResult] = React.useState<DataStoreCollectionResult<M>>({
-    items: [],
-    isLoading: false,
-    error: undefined,
-  });
+function useDataStoreResult<Model extends PersistentModel>(
+  props:
+    | DataStoreBindingProps<Model, 'record'>
+    | DataStoreBindingProps<Model, 'collection'>
+): DataStoreItemResult<Model> | DataStoreCollectionResult<Model> {
+  const isRecord = props.type === 'record';
 
-  const fetch = () => {
-    setResult({ isLoading: true, items: [] });
+  const [isLoading, setIsLoading] =
+    React.useState<
+      (
+        | DataStoreItemResult<Model>
+        | DataStoreCollectionResult<Model>
+      )['isLoading']
+    >(false);
+  const [error, setError] = React.useState<Error>();
+  const [items, setItems] = React.useState<
+    // use DataStoreCollectionResult for both 'record' and 'collection' types
+    DataStoreCollectionResult<Model>['items']
+  >([]);
+
+  const hasRan = React.useRef(false);
+  React.useEffect(() => {
+    if (hasRan.current) {
+      return;
+    }
+
+    setIsLoading(true);
+    // set to true to prevent additional Datastore call on re-render
+    hasRan.current = true;
+
+    if (isRecord) {
+      // call DataStore.query for single record
+      const { id, model } = props as any;
+      DataStore.query<Model>(model, id)
+        .then<void>((item) => {
+          setItems([item]);
+        })
+        .catch((error) => {
+          setError(error);
+        })
+        .finally(() => {
+          setIsLoading(false);
+        });
+    } else {
+      // call DataStore.observeQuery for record collection
+      const { criteria, model, pagination } = props;
 
-    const subscription = DataStore.observeQuery(
-      model,
-      criteria,
-      pagination
-    ).subscribe(
-      (snapshot) => setResult({ items: snapshot.items, isLoading: false }),
-      (error) => setResult({ items: [], error, isLoading: false })
-    );
+      const subscription = DataStore.observeQuery(
+        model,
+        criteria,
+        pagination
+      ).subscribe(
+        ({ items }) => {
+          setItems(items);
+          setIsLoading(false);
+        },
+        (error) => {
+          setError(error);
+          setIsLoading(false);
+        }
+      );
 
-    // Unsubscribe from query updates on unmount
-    if (subscription) {
-      return () => subscription.unsubscribe();
+      // Unsubscribe from query updates on unmount
+      if (subscription) {
+        return () => {
+          subscription.unsubscribe();
+        };
+      }
     }
-  };
+  }, [isRecord, props]);
```

Comment:
I like the solution of using `useRef` to prevent additional datastore calls. I think the infinite `useEffect` loop will arise again. People will normally pass in a new object `useDataStoreBinding({})` when they using the hook.

Reply:
Maybe I'm missing context here but not sure what the concern fully is, a `useEffect` with an empty dep array still gets called on each render, it just doesn't execute its body (unless I am mistaken)

Reply:
That's true from js perspective. I don't know how React emit this errors internally but this will arise when dependencies are changing constantly. They use a term `prevent re-running` on `useEffect` to explain the effect passing an empty array. It is close to the familiar `componentDidMount` and `componentWillUnmount` mental model.

Reply:
A potential solution would be to assign props to a ref which would allow us to remove it from the dep array, which would then prevent the useEffect from running unless isRecord updates. Thoughts?

Reply:
I like it. That's a way recommended by React actually. <- https://reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often

Reply:
Per offline discussion moving the refactoring of `useDataStoreCollection` and `useDataStoreItem` to their own PR